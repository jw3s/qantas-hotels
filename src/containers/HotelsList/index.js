import React, { Component } from 'react';
import Spinner from 'react-loader-spinner';
import { connect } from 'react-redux';
import styled from 'styled-components';
import {
  Card, Typography, AmountLabel, Ratings,
} from '../../components';
import type { Hotels } from '../../types/hotels';

import { getHotels } from '../../ducks/hotels';

const CardInfo = styled.div`
  flex: 1 0 0;
  display: flex;
  flex-direction: column;
  min-width: 400px;
  justify-content: space-between;
`;
const CardInfoContainer = styled.div`
  display: flex;
  flex-direction: column;
`;
const CardDetails = styled.div`
  flex: 0.25 0 0;
  display: flex;
  flex-wrap: wrap;
  @media (max-width: 619px) {
    margin-top: 40px;
  }
`;
const CardExtended = styled(Card)`
  &::before {
    position: absolute;
    content: '${({ content }) => content}';
    padding: 7px 20px;
    font-weight: 500;
    color: white;
    font-size: 12px;
    border-top-right-radius: 5px;
    border-bottom-right-radius: 5px;
    background: #cd0000;
    top: 10px;
    left: 0;
  }
`;

type Props = {
  getHotels: Function,
  hotels: Array<Hotels>,
  isLoading: boolean,
};

class HotelsList extends Component<Props> {
  componentDidMount() {
    const { getHotels: get } = this.props;
    get();
  }

  render() {
    const { hotels, isLoading } = this.props;
    if (isLoading) {
      return <Spinner type="Bars" color="#cd0000" height={80} width={80} />;
    }
    return (
      <div>
        {hotels.map(
          ({
            id,
            image,
            title,
            address,
            price,
            savings,
            rating,
            ratingType,
            freeCancellation,
            promotion,
            ...data
          }) => (
            <CardExtended key={id} image={image} content={promotion}>
              <CardInfo>
                <CardInfoContainer>
                  <Typography.Heading as="h2" mb={5}>
                    {title}
                  </Typography.Heading>
                  <Typography.Label size={12} mb={20}>
                    {address}
                  </Typography.Label>
                  <Typography.Label weight={500} size={12}>
                    {data.name || data.roomName}
                  </Typography.Label>
                </CardInfoContainer>
                {freeCancellation && (
                  <Typography.Label size={13} color="#1c6d4a">
                    Free cancellation
                  </Typography.Label>
                )}
              </CardInfo>
              <CardDetails>
                <Ratings value={rating} type={ratingType === 'self' ? 'circle' : 'star'} />
                <div>
                  <Typography.Label size={11}>
                    <b>1</b>
                    {' '}
night total in (AUD)
                  </Typography.Label>
                  <AmountLabel>{price}</AmountLabel>
                  {savings > 0 && (
                    <Typography.Label color="#fb0101" size={13} weight="700">
                      {`You save $${savings}~`}
                    </Typography.Label>
                  )}
                </div>
              </CardDetails>
            </CardExtended>
          ),
        )}
      </div>
    );
  }
}

export default connect(
  ({ hotels: { all, isLoading } }) => ({ hotels: all, isLoading }),
  { getHotels },
)(HotelsList);
