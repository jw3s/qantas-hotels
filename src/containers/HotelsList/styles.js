import styled from 'styled-components';
import { Card } from '../../components';

export const CardInfo = styled.div`
  flex: 1 0 0;
  display: flex;
  flex-direction: column;
  min-width: 400px;
  justify-content: space-between;
`;
export const CardInfoContainer = styled.div`
  display: flex;
  flex-direction: column;
`;
export const CardDetails = styled.div`
  flex: 0.25 0 0;
  display: flex;
  flex-wrap: wrap;
  @media (max-width: 619px) {
    margin-top: 40px;
  }
`;
export const CardExtended = styled(Card)`
  &::before {
    position: absolute;
    content: '${({ content }) => content}';
    padding: 7px 20px;
    font-weight: 500;
    color: white;
    font-size: 12px;
    border-top-right-radius: 5px;
    border-bottom-right-radius: 5px;
    background: #cd0000;
    top: 10px;
    left: 0;
    z-index: 1;
  }
`;
