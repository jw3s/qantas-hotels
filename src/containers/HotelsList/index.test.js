import React from 'react';
import { shallow } from 'enzyme';
import { createMockStore } from 'redux-test-utils';

import HotelsList from '.';

import 'jest-styled-components';

describe('<HotelsList />', () => {
  it('<HotelsList /> should render tree', () => {
    const testState = {
      hotels: { all: [0, 1, 2, 3], isLoading: false },
    };
    const store = createMockStore(testState);

    const tree = shallow(<HotelsList />, { context: { store } });
    expect(tree).toMatchSnapshot();
  });
  it('<HotelsList /> should render spinner if loading is true', () => {
    const testState = {
      hotels: { all: [0, 1, 2, 3], isLoading: true },
    };
    const store = createMockStore(testState);

    const tree = shallow(<HotelsList />, { context: { store } });
    expect(tree).toMatchSnapshot();
  });
});
