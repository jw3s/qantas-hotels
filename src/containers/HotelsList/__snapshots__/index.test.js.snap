// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HotelsList /> <HotelsList /> should render spinner if loading is true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(HotelsList) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "getHotels": [Function],
      "hotels": Array [
        0,
        1,
        2,
        3,
      ],
      "isLoading": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "getHotels": [Function],
        "hotels": Array [
          0,
          1,
          2,
          3,
        ],
        "isLoading": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;

exports[`<HotelsList /> <HotelsList /> should render tree 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(HotelsList) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "getHotels": [Function],
      "hotels": Array [
        0,
        1,
        2,
        3,
      ],
      "isLoading": false,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "getHotels": [Function],
        "hotels": Array [
          0,
          1,
          2,
          3,
        ],
        "isLoading": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
